### ISSUE TITLE
Support for palette rendering of SF symbols

### ISSUE THREAD

**Issue by bartreardon** [bartreardon/swiftDialog#131]
![image](https://user-images.githubusercontent.com/3598965/170847869-6a5c6f07-bb81-4e13-aa5b-fb6067af43c8.png)

![image](https://user-images.githubusercontent.com/3598965/170847874-00133f61-c02d-42fa-83c6-47e823290b5a.png)

**Comment by bartreardon** [comment]
@smithjw how is this implementation?
`--icon SF=person.3.sequence.fill,palette=red,palette2=green,palette3=blue`

![image](https://user-images.githubusercontent.com/3598965/170913980-e38aecb4-f9f1-47ca-8f64-2cf8ce4b53e0.png)

I could reuse the `colour` keyword but that should still work for gradients. the palette keyword I think fits the use better. on macOS 11 it will just revert to doing a gradient using the first two colours.

**Comment by smithjw** [comment]
Yeah I think it makes sense to use the `palette` keyword here. For icons, an admin would be looking at the following I think:
| Colour Type | Syntax |
| ----------- | ----------- |
| Multicolor      | `colour=auto`       |
| Hierarchical   | `colour=#4175C4`       |
| Gradient   | `colour1=#004165,colour2=#394A58`        |
| Palette   | `palette=red,palette2=green,palette3=blue`       |

Is it correct that swiftDialog uses hierarchical colour when just defining one vs monochrome?

Monochrome
![image](https://user-images.githubusercontent.com/916389/170915027-02fe6df5-add0-4a2c-8d77-707f07020486.png)

Hierarchical
![image](https://user-images.githubusercontent.com/916389/170915053-335a02e4-b1dc-4320-a164-3a319a75407b.png)

Multicolour
![image](https://user-images.githubusercontent.com/916389/170915092-e058edd0-4d5b-424c-b617-801b2db16766.png)

**Comment by bartreardon** [comment]
under macos 12 yes, it will use hierarchical when using a single colour.

thinking about the syntax, I think `palette=red,green,blue` vs `palette2=, palette3=` etc might be better. will see how that works.

**Comment by bartreardon** [comment]
re hierarchical vs mono - that's just the rendering mode. I could make it an optional but opted to default to hierarchical on macos 12 as it generally looks a lot nicer IMO.

**Comment by smithjw** [comment]
That's a lot easier to write so long as you just ignore any extra colours. Might also work a bit better for gradient colours too...
`colour=red,green` or `colour1=#004165,#394A58`

**Comment by bartreardon** [comment]
there are a lot of ways they can be styled that's for sure - you can even do gradients on individual palette items. I think that's a lot to try and stuff into a command argument structure though. Already pretty complex as it is.

**Comment by bartreardon** [comment]
`--icon SF=person.3.sequence.fill,palette=pink:orange:purple` 

what's a better separator than `:`? can't use `,` as that's being used to seperate the keywords in the argument

**Comment by bradtchapman** [comment]
I vote for using "space" as the separator and the = equals sign as the connector EXCEPT when it comes to the palette modes.  See my thoughts below.

As precedent, the 'log' command with the --predicate option already uses spaces to delineate operators, comparators, and keywords.  Require all "SF" options to be encapsulated in quotation marks or else fail it as unrecognized syntax.  

The `render` parameter does not need to be specified; however, if the user specifies 2+ colors without a `render` parameter, assume they want "palette" not gradient.  Most icons only have 1 color; some have 2.  Very few have 3.  If they intended for a gradient, make them write `render=gradient`.  

If you check the Color Picker and the SF Symbols apps, all of the 48 crayons and all 20 options, respectively, are **single words** with no spaces.  Thus, we do not need to worry about 'light green' or 'navy blue' as these do not exist in the standard 48.  If a user desires an exact shade, they need to use the #HHEEXX codes.

swiftDialog users should be advised of the following:

* The basic Swift colors change *slightly* when using dark vs light bg.
* The color options are only available on Monterey and up; Big Sur is limited to monochrome.
* Please [download](https://developer.apple.com/sf-symbols/) the SF Symbols browser if unsure about how colors will appear.

### Render Style 1: Monochrome Rendering (single color)

* Allow the use of "m" or "mono" as shorthand.
* If no color specified, assume black in light mode and white in dark mode
* When color is specified, accept the Apple palette name, the "crayon" name (in Color Picker), or the hex code as per usual.

```
--icon "SF=icon.name.variant color=gray"
--icon "SF=icon.name.variant color=#808080"
--icon "SF=icon.name.variant render=m,color=gray"
--icon "SF=icon.name.variant render=mono,color=gray bg=default|light|dark|maraschino|#ff2600"
```

### Render Style 2: Hierarchical (Ordered) Colors.

Personally, I think Apple is bonkers for using such an academic word to describe the ordered arrangement of colors on a an array of symbols.  For the sake of those implementing swiftDialog, I would suggest the use of "ordered" as it is both a synonym of hierarchical, and also much easier to explain and **translate** to other languages.  Assume the base color is black (light mode) or white (dark mode) unless a color is specified.  Both color and bg are optional as before.

Allow the use of "render=h" or "render=o" as shorthand.

```
--icon "SF=icon.name.variant render=h"
--icon "SF=icon.name.variant render=o color=gray"
--icon "SF=icon.name.variant render=ordered color=#808080"
--icon "SF=icon.name.variant render=ordered bg=#123456"
```

### Render Style 3: Palette 

* Allow the use of "p" or "pal" as shortcut for "palette."
* If a user only selects the first color, use the default 'secondary' and 'tertiary' colors.
* Accept up to 3 colors as input.
* *Four shalt thou not count, neither countest thou two excepting that thou proceedest directly to three.*
* *Five is right out!*
* Consider this variation: `"color=first color=second color=third"`
* Consider this variation: `"colors=first,second,third"`
* Consider alternate spellings of colour and colours.

```
--icon "SF=icon.name.variant render=palette color=green color=honeydew color=#4f8f00"
--icon "SF=icon.name.variant render=pal colours=green,honeydew,#4f8f00"
--icon "SF=icon.name.variant render=p colors=bubblegum,maroon bg=dark"
```

### Render Style 4: Multicolor

* Allow the shorthand "mc" or "multi" for "multicolor," but don't allow "m" (reserved for 'mono').
* No color needs to be specified here, as the symbols already have defined color choices per Apple.

```
--icon "SF=icon.name.variant render=mc"
--icon "SF=icon.name.variant render=multi color=indigo"
--icon "SF=icon.name.variant render=multicolor bg=dark"
--icon "SF=icon.name.variant render=mc color=#ff0000 bg=dark"
```

Render Style 5: Gradient

* This will create a diagonal gradient from top left to bottom right with two colors.
* Allow the shorthand "g" or "grad" for `render=gradient`

```
--icon "SF=icon.name.variant render=g colors=indigo,white"
--icon "SF=icon.name.variant render=gradient colors=#ccbb55,#123456"
```

**Comment by bartreardon** [comment]
@bradtchapman Thanks for the detailed feedback - The only thing I don't like doing is introducing breaking changes, so replacing `,` for space as the separators would introduce a breaking change for anyone using the current form and would also require enforced quotes. I do like the `renderer=` for forcing different rendering types though but I'd stick to a long/short form. It will allow me to seperate the logic a bit easier though I think.

**Comment by bartreardon** [comment]
ok - I think I have a nice way to handle the text input using the existing comma separated syntax but also allow for something like `palette=red,green,blue` . Will post a build later today.

**Comment by bartreardon** [comment]
check it

![image](https://user-images.githubusercontent.com/3598965/171320653-335fde05-90cf-494a-8764-39f856d8b480.png)

**Comment by bartreardon** [comment]
available in [1.11](https://github.com/bartreardon/swiftDialog/releases/tag/v1.11.0)
